<article class="day-desc"><h2>--- Day 6: Trash Compactor ---</h2><p>After helping the Elves in the kitchen, you were taking a break and helping them re-enact a movie scene when you over-enthusiastically jumped into the garbage chute!</p>
<p>A brief fall later, you find yourself in a <span title="To your surprise, the smell isn't that bad.">garbage smasher</span>. Unfortunately, the door's been magnetically sealed.</p>
<p>As you try to find a way out, you are approached by a family of cephalopods! They're pretty sure they can get the door open, but it will take some time. While you wait, they're curious if you can help the youngest cephalopod with her <a href="/2021/day/18">math homework</a>.</p>
<p>Cephalopod math doesn't look that different from normal math. The math worksheet (your puzzle input) consists of a list of <em>problems</em>; each problem has a group of numbers that need to be either <em>added</em> (<code>+</code>) or <em>multiplied</em> (<code>*</code>) together.</p>
<p>However, the problems are arranged a little strangely; they seem to be presented next to each other in a very long horizontal list. For example:</p>
<pre><code>123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
</code></pre>
<p>Each problem's numbers are arranged vertically; at the bottom of the problem is the symbol for the operation that needs to be performed. Problems are separated by a full column of only spaces. The left/right alignment of numbers within each problem can be ignored.</p>
<p>So, this worksheet contains four problems:</p>
<ul>
<li><code>123</code> * <code>45</code> * <code>6</code> = <code><em>33210</em></code></li>
<li><code>328</code> + <code>64</code> + <code>98</code> = <code><em>490</em></code></li>
<li><code>51</code> * <code>387</code> * <code>215</code> = <code><em>4243455</em></code></li>
<li><code>64</code> + <code>23</code> + <code>314</code> = <code><em>401</em></code></li>
</ul>
<p>To check their work, cephalopod students are given the <em>grand total</em> of adding together all of the answers to the individual problems. In this worksheet, the grand total is <code>33210</code> + <code>490</code> + <code>4243455</code> + <code>401</code> = <code><em>4277556</em></code>.</p>
<p>Of course, the actual worksheet is <em>much</em> wider. You'll need to make sure to unroll it completely so that you can read the problems clearly.</p>
<p>Solve the problems on the math worksheet. <em>What is the grand total found by adding together all of the answers to the individual problems?</em></p>
</article>

## Read Input

In [70]:
with open("./inputs/day06.txt") as f:
    input_string = f.read()

## Parse Input

In [71]:
def parse(text: str) -> list[tuple[str, list[int]]]:
    lines = text.strip().splitlines()
    split_lines = [line.split() for line in lines]
    problems: list[tuple[str, list[int]]] = []
    for i in range(len(split_lines[0])):
        operation = split_lines[-1][i]
        operands = [int(line[i]) for line in split_lines[:-1]]
        problems.append((operation, operands))
    return problems

data = parse(input_string)
print(data)

[('+', [5, 5, 61, 62]), ('*', [679, 847, 86, 3]), ('*', [21, 87, 49, 73]), ('+', [1766, 277, 467, 51]), ('+', [81, 81, 38, 77]), ('*', [75, 51, 211, 483]), ('+', [16, 62, 37, 23]), ('+', [993, 98, 55, 49]), ('+', [2897, 823, 3, 8]), ('*', [28, 64, 38, 5]), ('+', [534, 254, 945, 3]), ('+', [9545, 144, 644, 5]), ('*', [5, 37, 94, 95]), ('+', [5922, 3422, 394, 634]), ('*', [63, 1, 3, 3]), ('+', [6, 1, 52, 77]), ('*', [732, 636, 535, 379]), ('*', [569, 778, 977, 915]), ('*', [579, 546, 923, 337]), ('*', [727, 317, 875, 714]), ('+', [63, 98, 24, 56]), ('+', [57, 82, 8, 8]), ('+', [931, 971, 837, 53]), ('+', [1854, 336, 262, 24]), ('*', [55, 89, 21, 1]), ('+', [834, 838, 64, 11]), ('*', [441, 153, 246, 39]), ('*', [32, 11, 51, 768]), ('+', [72, 291, 6193, 4979]), ('*', [33, 449, 847, 797]), ('+', [3, 3, 43, 485]), ('*', [365, 384, 62, 9]), ('+', [56, 9448, 4567, 9224]), ('+', [718, 511, 386, 7]), ('*', [96, 254, 848, 785]), ('*', [278, 118, 623, 576]), ('+', [985, 545, 9711, 9984]), ('*', [4

## Solve Part 1

In [72]:
def solve(problems: list[tuple[str, list[int]]]) -> int:
    total = 0
    for operation, operands in problems:
        if operation == '+':
            result = sum(operands)
        elif operation == '*':
            result = 1
            for op in operands:
                result *= op
        total += result
    return total

result = solve(data)
print(result)

4076006202939


<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The big cephalopods come back to check on how things are going. When they see that your grand total doesn't match the one expected by the worksheet, they realize they forgot to explain how to read cephalopod math.</p>
<p>Cephalopod math is written <em>right-to-left in columns</em>. Each number is given in its own column, with the most significant digit at the top and the least significant digit at the bottom. (Problems are still separated with a column consisting only of spaces, and the symbol at the bottom of the problem is still the operator to use.)</p>
<p>Here's the example worksheet again:</p>
<pre><code>123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
</code></pre>
<p>Reading the problems right-to-left one column at a time, the problems are now quite different:</p>
<ul>
<li>The rightmost problem is <code>4</code> + <code>431</code> + <code>623</code> = <code><em>1058</em></code></li>
<li>The second problem from the right is <code>175</code> * <code>581</code> * <code>32</code> = <code><em>3253600</em></code></li>
<li>The third problem from the right is <code>8</code> + <code>248</code> + <code>369</code> = <code><em>625</em></code></li>
<li>Finally, the leftmost problem is <code>356</code> * <code>24</code> * <code>1</code> = <code><em>8544</em></code></li>
</ul>
<p>Now, the grand total is <code>1058</code> + <code>3253600</code> + <code>625</code> + <code>8544</code> = <code><em>3263827</em></code>.</p>
<p>Solve the problems on the math worksheet again. <em>What is the grand total found by adding together all of the answers to the individual problems?</em></p>
</article>

## Parse Part 2

In [76]:
def parse(text: str) -> list[tuple[str, list[int]]]:
    lines = text.splitlines()
    problems: list[tuple[str, list[int]]] = [(None, [])]

    for i in range(len(lines[0]) - 1, -1, -1):
        column = ''.join(line[i] for line in lines[:-1])
        if column.strip() != '':
            problems[-1][1].append(int(column))
        if lines[-1][i] != ' ':
            problems[-1] = (lines[-1][i], problems[-1][1])
            problems.append((None, []))

    return problems[:-1]
    

data_part2 = parse(input_string)
print(data_part2)

[('*', [2599, 373, 68]), ('+', [5174, 7]), ('+', [5289, 4712, 697]), ('+', [4947, 1432]), ('*', [311, 1715, 5259]), ('+', [4335, 632]), ('*', [2543, 9428]), ('*', [9256, 4226]), ('*', [5, 115, 7287]), ('*', [6718, 6514, 2811]), ('*', [1491, 552, 74]), ('+', [9239, 8318, 8334, 1192]), ('+', [5722, 4175]), ('*', [2559, 9626]), ('*', [88, 6469, 3465]), ('*', [3733, 929, 5]), ('*', [2564, 1363]), ('+', [9566, 767, 91, 7]), ('*', [2725, 8475]), ('+', [7927, 9574, 26]), ('+', [24, 9487, 4876]), ('*', [2165, 681]), ('*', [5494, 924]), ('*', [3562, 198]), ('*', [69, 757, 7476]), ('*', [1218, 3985]), ('*', [2455, 4312, 672]), ('+', [1123, 4959, 8927, 821]), ('+', [8344, 3552]), ('+', [7612, 7489, 3922]), ('+', [76, 8396]), ('*', [21, 194, 5939]), ('*', [226, 167, 6394]), ('*', [2433, 73]), ('*', [563, 8168, 2539]), ('*', [7, 718, 6588]), ('*', [3965, 28]), ('+', [2525, 345, 31, 54]), ('+', [4118, 4574, 13]), ('+', [246, 2729, 9163]), ('*', [9764, 815]), ('+', [76, 8782]), ('+', [89, 9168]), ('*

## Solve Part 2

In [77]:
result = solve(data_part2)
print(result)

7903168391557
